Custom pronouns & presets

DeadGirlRisen opened the pull request:
Alright, with this we should have a custom pronouns screen so you can customize which pronouns a cat goes by, and can add in your own pronouns.  We've worked hard on this, and hopefully now this time we got everything ready.

DeadGirlRisen committed:
Pronouns, correct

Custom Pronouns & Presets

DeadGirlRisen committed:
Revert "Pronouns, correct"

This reverts commit efafb7c0a874ddc301b2432161814d14b710cf36.

DeadGirlRisen committed:
Pronouns & Presets

This update allows cats to have pronouns, and customized pronouns.

DeadGirlRisen committed:
Shifting Window Over

DeadGirlRisen committed:
removing window

DeadGirlRisen committed:
Update

DeadGirlRisen committed:
Update clan.py

Saving Custom Pronouns for clan specific

DeadGirlRisen committed:
Merge branch 'ClanGenOfficial:development' into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
Merge branch 'ClanGenOfficial:development' into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
new buttons

DeadGirlRisen committed:
Merge branch 'ClanGenOfficial:development' into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
Merge branch 'ClanGenOfficial:development' into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
UI Update

DeadGirlRisen committed:
Merge branch 'Custom-Pronouns-&-Presets' of https://github.com/DeadGirlRisen/Lunar-Mod into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
Merge branch 'ClanGenOfficial:development' into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
Minor fix

DeadGirlRisen committed:
Merge branch 'ClanGenOfficial:development' into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
Minor fixes

DeadGirlRisen committed:
Merge branch 'Custom-Pronouns-&-Presets' of https://github.com/DeadGirlRisen/Lunar-Mod into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
Pronoun Game Settings

DeadGirlRisen committed:
gender identity location

DeadGirlRisen committed:
Gender Align Fix

DeadGirlRisen committed:
Fixing Gender Align

DeadGirlRisen committed:
Fix Trans Baby bug

DeadGirlRisen committed:
Update ChangeGenderScreen.py

DeadGirlRisen committed:
Update cats.py

DeadGirlRisen committed:
Fullscreen Fix

scribblecrumb committed:
If cat has no gender_align, display gender instead

scribblecrumb committed:
adding tooltips and small UI adjustments

DeadGirlRisen committed:
Merge pull request #5 from scribblecrumb/Custom-Pronouns-&-Presets

Some small adjustments

Commit:
Merge branch 'Custom-Pronouns-&-Presets' of https://github.com/DeadGirlRisen/Lunar-Mod into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
Renaming

DeadGirlRisen committed:
Update tool_tips.json

DeadGirlRisen committed:
Update windows.py

DeadGirlRisen committed:
Merge branch 'ClanGenOfficial:development' into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
Merge branch 'ClanGenOfficial:development' into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
minor changes

2 & 6 done

DeadGirlRisen committed:
Merge branch 'Custom-Pronouns-&-Presets' of https://github.com/DeadGirlRisen/Lunar-Mod into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
Update cats.py

#7 done

DeadGirlRisen committed:
Update cats.py

DeadGirlRisen committed:
5,8&Bug

5) Save only as preset fixed
8) Checkbox renaming done.

Bug Fix: Trans cats now get proper pronouns when transing via an event.

DeadGirlRisen committed:
Merge branch 'ClanGenOfficial:development' into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
Update ProfileScreen.py

KiraDragoness committed:
Update general.json

DeadGirlRisen committed:
Merge branch 'ClanGenOfficial:development' into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
Merge branch 'ClanGenOfficial:development' into Custom-Pronouns-&-Presets

KiraDragoness committed:
Update general.json

DeadGirlRisen committed:
Merge pull request #8 from KiraDragoness/Custom-Pronouns-&-Presets

Pronoun tagging in general thought.json

KiraDragoness committed:
Fix pronoun taggings

KiraDragoness committed:
Update clan_reaction.json

KiraDragoness committed:
Update training.json

DeadGirlRisen committed:
Update clan_cat.py

DeadGirlRisen committed:
Merge pull request #9 from KiraDragoness/Custom-Pronouns-&-Presets

Fix pronoun taggings

KiraDragoness committed:
Fixed some patrol pronoun taggings

KiraDragoness committed:
Update any.json

DeadGirlRisen committed:
Merge pull request #10 from KiraDragoness/Custom-Pronouns-&-Presets

Fixed some patrol pronoun taggings

KiraDragoness committed:
Patrol pronoun tags fixing

DeadGirlRisen committed:
Merge branch 'Custom-Pronouns-&-Presets' into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
Merge pull request #11 from KiraDragoness/Custom-Pronouns-&-Presets

Patrol pronoun tags fixing

DeadGirlRisen committed:
Merge branch 'Custom-Pronouns-&-Presets' of https://github.com/DeadGirlRisen/Lunar-Mod into Custom-Pronouns-&-Presets

Commit:
Merge branch 'Custom-Pronouns-&-Presets' of https://github.com/DeadGirlRisen/Lunar-Mod into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
Updated Files

Updated files from Bingle

DeadGirlRisen committed:
Merge branch 'Custom-Pronouns-&-Presets' of https://github.com/DeadGirlRisen/Lunar-Mod into Custom-Pronouns-&-Presets

DeadGirlRisen committed:
Update all_screens.py

keyraven commented:
I will do a full review of this later. From a quick functionality check, the game currently crashes when pressing the"x" button in the "Add Pronoun" dialogue.

keyraven commented on resources/dicts/pronouns.json:
```diff
@@ -0,0 +1,31 @@
+{
+  "default_pronouns": [
+    {
+      "ID": "Neutral",
```
As I've mentioned in previous reviews, I do not see the need for the "ID" field for pronouns. You can check for strict equality of the entire pronoun dictionary for deletion/identification. The "ID" field is not shown to the player, and adds further complexity.

DeadGirlRisen replied:
Bingle and I would like help with this 💜

DeadGirlRisen commented:
> I will do a full review of this later. From a quick functionality check, the game currently crashes when pressing the"x" button in the "Add Pronoun" dialogue.

Thats odd, I'm not having that happen myself.

Wait, it happens after you save a pronoun oddly enough, but not if you don't add a pronoun. I'll take a look, somehow in the testing that bug was not encountered before? Or it might be due to the update from bingle. Thanks for finding it, I'll see what I can do.

DeadGirlRisen committed:
Update windows.py

Fixes the crash after saving a pronoun and hitting the exit button.

DeadGirlRisen commented:
> I will do a full review of this later. From a quick functionality check, the game currently crashes when pressing the"x" button in the "Add Pronoun" dialogue.

Alright, it looks like I fixed it from a quick test on my end, it was just a matter of one of the bits of code forgot to be changed over with the new changes bingle made.

keyraven commented on resources/dicts/events/ceremonies/ceremony-master.json:
```diff
@@ -2056,7 +2056,7 @@
       "all_traits"
     ],
-    "l_n calls the Clan to a meeting and declares (old_name) to be a warrior. {PRONOUN/m_c/subject/CAP} {VERB/m_c/are/is} now called m_c and are celebrated for {PRONOUN/m_c/poss} r_h."
+    "l_n calls the Clan to a meeting and declares (old_name) to be a warrior. {PRONOUN/m_c/subject/CAP} {VERB/m_c/are/is} now called m_c and {VERB/m_c/are/is} celebrated for {PRONOUN/m_c/poss} r_h."
```
This is a large pull request, so it would make it easier to review if these non-related pronoun tag corrections were put in a separate pull request. At this point, it may be too much work. But in the future, it would be best to keep large feature addition PRs focused on the feature added.

DeadGirlRisen replied:
Will do that in the future 👍

keyraven commented on scripts/cat/cats.py:
```diff
@@ -280,66 +283,91 @@ def __init__(self,
+            if self.age in ['kitten', 'newborn']:
+                trans_chance = 0
+                nb_chance = 0
+            if game:
```
If, for whatever reason, "game" is falsy, a cat will generate with an empty pronoun list, which will crash the game when text replacement needs to happen. Rather than putting the entire pronoun determination inside the "if game" check, set a local variable with the value of "they them default", then proceed with pronoun determination.

DeadGirlRisen replied:
Marking as fixed

keyraven commented on scripts/cat/cats.py:
```diff
@@ -280,66 +283,91 @@ def __init__(self,
-            else:
-                self.experience = 0
+        #Personality
+        self.personality = Personality(kit_trait=self.is_baby())
```
This section should remain indented.  It needs to remain under the current if-statement.

DeadGirlRisen replied:
I believe I indented it correctly now :D

keyraven commented on scripts/cat/cats.py:
```diff
@@ -1890,7 +1918,7 @@ def contact_with_ill_cat(self, cat: Cat):
                 for y in self.injuries:
                     illness_infect = list(
-                        filter(lambda ill: ill["name"] == illness_name, self.injuries[y]["illness_infectiousness"]))
+                        filter(lambda ill: ill["ID"] == illness_name, self.injuries[y]["illness_infectiousness"]))
```
What is the purpose of this change? It doesn't seem to have anything to do w/ pronouns. Is this an unrelated bug fix?

DeadGirlRisen replied:
My mistake has been fixed.

keyraven commented on scripts/cat/cats.py:
```diff
@@ -280,66 +283,91 @@ def __init__(self,
+                        else:
+                            self.genderalign = "trans female"
+                #assigning pronouns to afab cats            
+                elif self.gender == "female":
```
The section for AFAB and AMAB cats is also identical. You could improve readability by condensing into one section.  Or (even better), separating out gender and pronoun assignment so that gender is determined first, then pronouns.

DeadGirlRisen replied:
Fixed! I believe I streamlined it, and it looked like it worked in my testing.

keyraven commented on scripts/cat/cats.py:
```diff
@@ -280,66 +283,91 @@ def __init__(self,
                 
-            if not skill_dict:
-                self.skills = CatSkills.generate_new_catskills(self.status, self.moons)
+        if not skill_dict:
```
Same here - re-indent this

DeadGirlRisen replied:
Reindented! :D

keyraven commented on scripts/clan.py:
```diff
@@ -457,6 +463,9 @@ def save_clan(self):
         clan_data["med_cat_predecessors"] = self.med_cat_predecessors
+        
+        if self.custom_pronouns:
+            clan_data["custom_pronouns"] = self.custom_pronouns
```
There is no need for lines 467 and 468. self.custom_pronouns is already assigned to the key "custom_pronouns" on line 439.

DeadGirlRisen replied:
Deleted 👍

keyraven commented on scripts/clan.py:
```diff
@@ -746,6 +755,11 @@ def load_clan_json(self):
+            if clan_data["custom_pronouns"]:
+                game.clan.custom_pronouns = clan_data["custom_pronouns"]
+        else:
+            game.clan.custom_pronouns = []
```
I would remove this "else" statement, since custom_pronouns is already set as a empty list in the clan constructor.

DeadGirlRisen replied:
fixed :D

keyraven commented on scripts/game_structure/load_cat.py:
```diff
@@ -121,6 +121,8 @@ def json_load():
             
             new_cat.genderalign = cat["gender_align"]
+            # new_cat.pronouns = cat["pronouns"]
+            new_cat.pronouns = cat["pronouns"] if "pronouns" in cat else [new_cat.default_pronouns[0].copy()]
```
Note -- cats from old saves will be assigned they/them pronouns irrregardless of the setting. This is probably fine, but important to note.

keyraven commented on scripts/game_structure/load_cat.py:
```diff
@@ -182,6 +184,7 @@ def json_load():
                 )
 
             # new_cat.pronouns = cat["pronouns"] if "pronouns" in cat else [new_cat.default_pronouns[0].copy()]
+            # new_cat.pronouns = cat.get("pronouns", new_cat.default_pronouns[0].copy())
```
If these lines are no longer necessary, and replaced by line 125, could you delete them to keep the loading code tidy? Same with line 124.

DeadGirlRisen replied:
Done :D

keyraven commented on scripts/screens/ChangeGenderScreen.py:
```diff
@@ -0,0 +1,631 @@
+                self.next_cat_button.disable()
+                self.back_button.disable()
+
+            else:
```
Having a "else" statement for button event handling is confusing, and could possibly cause issues down the line.  It took me a while to figure out what this section of code was for.   If would be clearer to use additional elif statements, rather than a blanket "else" statement for button clicks.

DeadGirlRisen replied:
I believe I fixed this, though I think when you fix the way the pronouns are identified this area of code might be changed as well.

keyraven commented on scripts/screens/ChangeGenderScreen.py:
```diff
@@ -0,0 +1,631 @@
+
+                for pronounset in pronouns_data["default_pronouns"]:
+                    add_button_id = f"add_button_{pronounset['ID']}" + "preset"
+                    if event.ui_element == self.buttons.get(add_button_id):
```
Rather than looping button keys, I would suggest using something similar to CatButton or UISpriteButton, which adds an additional data field to the button object, which can then be extracted from the button event. In fact, you would probably add an object_id field to the CatButton constructor and use that class directly.  This would also help remove the need for the 'ID' value for pronoun dictionaries.

DeadGirlRisen replied:
I believe this is part of what you're helping with, so I leave it in your capable hands  :D

keyraven commented on scripts/screens/ChangeGenderScreen.py:
```diff
@@ -0,0 +1,631 @@
+                        self.update_selected_cat()
+                        break
+
+        elif event.type == pygame.KEYDOWN and game.settings['keybinds']:
```
Probably best to remove keybinds for now. They are a pain and a half. As this code is currently set-up, it would allow game.switches['cat'] to be set to an invalid cat ID.

DeadGirlRisen replied:
Removed :D

keyraven commented on scripts/screens/ProfileScreen.py:
```diff
@@ -263,12 +263,12 @@ def handle_tab_events(self, event):
-                SpecifyCatGender(self.the_cat)
-                '''if self.the_cat.genderalign in ["female", "trans female"]:
+                self.change_screen("change gender screen")
+                if self.the_cat.genderalign in ["female", "trans female"]:
```
This is currently the origin of a bug,where going to the cat's gender specify page will re-set their pronouns to correspond with their gender, clearing any previously set pronouns and ignoring the they/them default setting. Remove lines 266-271 to fix this issue.

DeadGirlRisen replied:
Whoops, good catch, I remember thinking to myself a few months back I need to fix that, guess I forgot XD

keyraven commented on scripts/utility.py:
```diff
@@ -1295,7 +1295,7 @@ def generate_sprite(cat, life_state=None, scars_hidden=False, acc_hidden=False,
         eyes = sprites.sprites['eyes' + cat.pelt.eye_colour + cat_sprite].copy()
         if cat.pelt.eye_colour2 != None:
-            eyes.blit(sprites.sprites['eyes2' + cat.pelt.eye_colour2 + cat_sprite], (0, 0))
+            eyes.blit(sprites.sprites["eyes2" + cat.pelt.eye_colour2 + cat_sprite], (0, 0))
```
What is the purpose of this change? Is this to fix an unrelated bug?  Same with the new "eye2color.png" file.

DeadGirlRisen replied:
Just a mistake when I was working on a different branch and hadn't properly changed my branches over. It should be fixed.

keyraven reviewed (changes requested):
It's looking much better! Mostly works great. I found a few issues, and had a few comments on the ChangeGenderScreen code.

DeadGirlRisen committed:
Fixes

Fixes some of the things mentioned by Key. I closed those parts of the conversations that I had done.

DeadGirlRisen committed:
Fixes part two

Fixed one mistake on my part where I deleted more than I needed, then did the rework on the else into an elif statement.

keyraven committed:
Refactor w/ CatButton

keyraven committed:
Merge branch 'development' into pr/2313

keyraven committed:
Remove ID Remains

keyraven reviewed (approved):
Looking Good!
